File: mockserver-client-go/expectation.go
package mockserver

// Expectation represents a MockServer expectation (request matcher + action).
type Expectation struct {
	ID       string `json:"id,omitempty"`
	Priority int    `json:"priority,omitempty"`
	// Percentage (0-100) makes the expectation match only that fraction of
	// otherwise-matching requests (probabilistic matching).
	Percentage int `json:"percentage,omitempty"`
	// Chaos applies a declarative fault-injection profile to the response.
	Chaos *HttpChaosProfile `json:"chaos,omitempty"`
	// RateLimit applies a declarative, protocol-agnostic rate limit / quota.
	RateLimit            *RateLimit      `json:"rateLimit,omitempty"`
	HttpRequest          *HttpRequest    `json:"httpRequest,omitempty"`
	HttpResponse         *HttpResponse   `json:"httpResponse,omitempty"`
	HttpResponses        []*HttpResponse `json:"httpResponses,omitempty"`
	HttpResponseTemplate *HttpTemplate   `json:"httpResponseTemplate,omitempty"`
	HttpForward          *HttpForward    `json:"httpForward,omitempty"`
	HttpForwardTemplate  *HttpTemplate   `json:"httpForwardTemplate,omitempty"`
	// HttpOverrideForwardedRequest forwards the matched request to an overridden
	// target (RequestOverride) while shaping the caller's response via
	// ResponseOverride or ResponseTemplate.
	HttpOverrideForwardedRequest *HttpOverrideForwardedRequest `json:"httpOverrideForwardedRequest,omitempty"`
	// HttpForwardWithFallback forwards upstream but falls back to a local
	// response on configured status codes or a timeout.
	HttpForwardWithFallback *HttpForwardWithFallback `json:"httpForwardWithFallback,omitempty"`
	// HttpForwardValidateAction forwards upstream and validates the request
	// and/or response against an OpenAPI spec.
	HttpForwardValidateAction *HttpForwardValidateAction `json:"httpForwardValidateAction,omitempty"`
	HttpError                 *HttpError                 `json:"httpError,omitempty"`
	// Class callbacks reference a server-side class (REST-only, no WebSocket).
	HttpResponseClassCallback *HttpClassCallback `json:"httpResponseClassCallback,omitempty"`
	HttpForwardClassCallback  *HttpClassCallback `json:"httpForwardClassCallback,omitempty"`
	// Object/closure callbacks are driven over the callback WebSocket; ClientId
	// identifies this client's WebSocket connection.
	HttpResponseObjectCallback *HttpObjectCallback    `json:"httpResponseObjectCallback,omitempty"`
	HttpForwardObjectCallback  *HttpObjectCallback    `json:"httpForwardObjectCallback,omitempty"`
	HttpSseResponse            *HttpSseResponse       `json:"httpSseResponse,omitempty"`
	HttpWebSocketResponse      *HttpWebSocketResponse `json:"httpWebSocketResponse,omitempty"`
	GrpcStreamResponse         *GrpcStreamResponse    `json:"grpcStreamResponse,omitempty"`
	// GrpcBidiResponse is a gRPC bidirectional-streaming response action.
	GrpcBidiResponse       *GrpcBidiResponse       `json:"grpcBidiResponse,omitempty"`
	BinaryResponse         *BinaryResponse         `json:"binaryResponse,omitempty"`
	DnsResponse            *DnsResponse            `json:"dnsResponse,omitempty"`
	HttpLlmResponse        *HttpLlmResponse        `json:"httpLlmResponse,omitempty"`
	ScenarioName           string                  `json:"scenarioName,omitempty"`
	ScenarioState          string                  `json:"scenarioState,omitempty"`
	NewScenarioState       string                  `json:"newScenarioState,omitempty"`
	ResponseMode           ResponseMode            `json:"responseMode,omitempty"`
	ResponseWeights        []int                   `json:"responseWeights,omitempty"`
	SwitchAfter            *int                    `json:"switchAfter,omitempty"`
	CrossProtocolScenarios []CrossProtocolScenario `json:"crossProtocolScenarios,omitempty"`
	// BeforeActions run before the primary action; AfterActions run after it
	// (e.g. fire a webhook or invoke a callback as a side effect).
	BeforeActions []AfterAction `json:"beforeActions,omitempty"`
	AfterActions  []AfterAction `json:"afterActions,omitempty"`
	// Steps defines a multi-step expectation (an ordered pipeline of actions).
	Steps []ExpectationStep `json:"steps,omitempty"`
	// Capture extracts values from the matched request into scenario/state keys.
	Capture []CaptureRule `json:"capture,omitempty"`
	// Namespace (tenant) isolates this expectation: when set it only matches
	// requests carrying the configured matchNamespaceHeader with this value.
	Namespace  string      `json:"namespace,omitempty"`
	Times      *Times      `json:"times,omitempty"`
	TimeToLive *TimeToLive `json:"timeToLive,omitempty"`
	// Timestamp is the server-assigned creation time of the expectation (echoed
	// back when the server returns registered expectations); ISO-8601 string.
	Timestamp string `json:"timestamp,omitempty"`
}

// ResponseMode selects how MockServer chooses between multiple HttpResponses
// (see Expectation.HttpResponses).
type ResponseMode string

const (
	// ResponseModeSequential cycles through the responses in order (the default).
	ResponseModeSequential ResponseMode = "SEQUENTIAL"
	// ResponseModeRandom picks a response at random.
	ResponseModeRandom ResponseMode = "RANDOM"
	// ResponseModeWeighted picks a response using ResponseWeights as relative weights.
	ResponseModeWeighted ResponseMode = "WEIGHTED"
	// ResponseModeSwitch advances to the next response after SwitchAfter requests.
	ResponseModeSwitch ResponseMode = "SWITCH"
)

// CrossProtocolTrigger identifies the kind of event that advances a scenario
// state machine across protocols.
type CrossProtocolTrigger string

const (
	CrossProtocolTriggerDNSQuery         CrossProtocolTrigger = "DNS_QUERY"
	CrossProtocolTriggerWebSocketConnect CrossProtocolTrigger = "WEBSOCKET_CONNECT"
	CrossProtocolTriggerGrpcRequest      CrossProtocolTrigger = "GRPC_REQUEST"
	CrossProtocolTriggerHTTPRequest      CrossProtocolTrigger = "HTTP_REQUEST"
)

// CrossProtocolScenario advances a named scenario to a target state when an
// event of the given Trigger kind is observed (optionally filtered by
// MatchPattern, a substring of the event identifier).
type CrossProtocolScenario struct {
	Trigger      CrossProtocolTrigger `json:"trigger"`
	MatchPattern string               `json:"matchPattern,omitempty"`
	ScenarioName string               `json:"scenarioName"`
	TargetState  string               `json:"targetState"`
}

// Times controls how many times an expectation can be matched.
type Times struct {
	RemainingTimes int  `json:"remainingTimes,omitempty"`
	Unlimited      bool `json:"unlimited"`
}

// TimeToLive controls how long an expectation remains active.
type TimeToLive struct {
	TimeUnit   string `json:"timeUnit,omitempty"`
	TimeToLive int    `json:"timeToLive,omitempty"`
	Unlimited  bool   `json:"unlimited"`
}

// Once returns a Times that matches exactly once.
func Once() *Times {
	return &Times{RemainingTimes: 1, Unlimited: false}
}

// Exactly returns a Times that matches exactly n times.
func Exactly(n int) *Times {
	return &Times{RemainingTimes: n, Unlimited: false}
}

// Unlimited returns a Times that matches unlimited times.
func Unlimited() *Times {
	return &Times{Unlimited: true}
}

// TTL returns a TimeToLive with the given duration.
func TTL(timeUnit string, value int) *TimeToLive {
	return &TimeToLive{TimeUnit: timeUnit, TimeToLive: value, Unlimited: false}
}

// UnlimitedTTL returns an unlimited TimeToLive.
func UnlimitedTTL() *TimeToLive {
	return &TimeToLive{Unlimited: true}
}

File: mockserver-client-dotnet/src/MockServer.Client/Models/Expectation.cs
using System.Text.Json.Serialization;
using MockServer.Client.Llm;

namespace MockServer.Client.Models;

/// <summary>
/// Represents a MockServer expectation (request matcher + action).
/// </summary>
public sealed class Expectation
{
    [JsonPropertyName("id")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Id { get; set; }

    [JsonPropertyName("priority")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Priority { get; set; }

    /// <summary>
    /// Probabilistic match percentage (0-100): when set, the expectation matches only this percentage
    /// of otherwise-matching requests. Drives simple probabilistic fault/behaviour injection.
    /// </summary>
    [JsonPropertyName("percentage")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Percentage { get; set; }

    /// <summary>
    /// Declarative HTTP chaos / fault-injection profile applied to this expectation.
    /// </summary>
    [JsonPropertyName("chaos")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpChaosProfile? Chaos { get; set; }

    /// <summary>
    /// Declarative, protocol-agnostic rate limit / quota applied to this expectation.
    /// </summary>
    [JsonPropertyName("rateLimit")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public RateLimit? RateLimit { get; set; }

    [JsonPropertyName("httpRequest")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpRequest? HttpRequest { get; set; }

    [JsonPropertyName("httpResponse")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpResponse? HttpResponse { get; set; }

    /// <summary>
    /// Multiple responses for a stateful scenario. When set, takes priority over the singular
    /// <see cref="HttpResponse"/>; the response served per match is selected by <see cref="ResponseMode"/>.
    /// </summary>
    [JsonPropertyName("httpResponses")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<HttpResponse>? HttpResponses { get; set; }

    [JsonPropertyName("httpForward")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpForward? HttpForward { get; set; }

    [JsonPropertyName("httpResponseTemplate")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpTemplate? HttpResponseTemplate { get; set; }

    [JsonPropertyName("httpForwardTemplate")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpTemplate? HttpForwardTemplate { get; set; }

    /// <summary>
    /// Override-forwarded-request action: the matched request is forwarded after applying a
    /// request override, while a response template produces the caller's response.
    /// </summary>
    [JsonPropertyName("httpOverrideForwardedRequest")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpOverrideForwardedRequest? HttpOverrideForwardedRequest { get; set; }

    /// <summary>
    /// Forward action that falls back to a canned response when the upstream forward fails or returns
    /// a configured status code.
    /// </summary>
    [JsonPropertyName("httpForwardWithFallback")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpForwardWithFallback? HttpForwardWithFallback { get; set; }

    /// <summary>
    /// Forward action that validates the request and/or response against an OpenAPI spec.
    /// </summary>
    [JsonPropertyName("httpForwardValidateAction")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpForwardValidateAction? HttpForwardValidateAction { get; set; }

    [JsonPropertyName("httpError")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpError? HttpError { get; set; }

    /// <summary>
    /// Response class-callback action: a server-side class produces the response.
    /// REST-only — no WebSocket involved.
    /// </summary>
    [JsonPropertyName("httpResponseClassCallback")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpClassCallback? HttpResponseClassCallback { get; set; }

    /// <summary>
    /// Forward class-callback action: a server-side class produces the request to forward.
    /// REST-only — no WebSocket involved.
    /// </summary>
    [JsonPropertyName("httpForwardClassCallback")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpClassCallback? HttpForwardClassCallback { get; set; }

    /// <summary>
    /// Response object-callback action: the request is dispatched over the callback
    /// WebSocket to the client identified by <c>clientId</c>, whose closure produces the response.
    /// </summary>
    [JsonPropertyName("httpResponseObjectCallback")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpObjectCallback? HttpResponseObjectCallback { get; set; }

    /// <summary>
    /// Forward object-callback action: the request is dispatched over the callback
    /// WebSocket to the client identified by <c>clientId</c>, whose closure produces the request to forward.
    /// </summary>
    [JsonPropertyName("httpForwardObjectCallback")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpObjectCallback? HttpForwardObjectCallback { get; set; }

    [JsonPropertyName("httpSseResponse")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpSseResponse? HttpSseResponse { get; set; }

    [JsonPropertyName("httpWebSocketResponse")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpWebSocketResponse? HttpWebSocketResponse { get; set; }

    [JsonPropertyName("grpcStreamResponse")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public GrpcStreamResponse? GrpcStreamResponse { get; set; }

    /// <summary>
    /// Bidirectional-streaming gRPC response action.
    /// </summary>
    [JsonPropertyName("grpcBidiResponse")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public GrpcBidiResponse? GrpcBidiResponse { get; set; }

    [JsonPropertyName("binaryResponse")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public BinaryResponse? BinaryResponse { get; set; }

    [JsonPropertyName("dnsResponse")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public DnsResponse? DnsResponse { get; set; }

    [JsonPropertyName("httpLlmResponse")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public HttpLlmResponse? HttpLlmResponse { get; set; }

    /// <summary>
    /// Side-effect actions run before the primary action produces a response.
    /// </summary>
    [JsonPropertyName("beforeActions")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<AfterAction>? BeforeActions { get; set; }

    /// <summary>
    /// Side-effect actions run after the primary action produces a response.
    /// </summary>
    [JsonPropertyName("afterActions")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<AfterAction>? AfterActions { get; set; }

    /// <summary>
    /// Ordered multi-step action sequence; exactly one step is the responder.
    /// </summary>
    [JsonPropertyName("steps")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<ExpectationStep>? Steps { get; set; }

    /// <summary>
    /// Value-capture rules that extract values from the matched request into scenario state.
    /// </summary>
    [JsonPropertyName("capture")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<CaptureRule>? Capture { get; set; }

    /// <summary>
    /// Optional namespace (tenant) this expectation belongs to; when set, the expectation only matches
    /// requests carrying the configured match-namespace header with this value and stays isolated from
    /// other namespaces.
    /// </summary>
    [JsonPropertyName("namespace")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Namespace { get; set; }

    [JsonPropertyName("scenarioName")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? ScenarioName { get; set; }

    [JsonPropertyName("scenarioState")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? ScenarioState { get; set; }

    [JsonPropertyName("newScenarioState")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? NewScenarioState { get; set; }

    /// <summary>
    /// How a response is selected from <see cref="HttpResponses"/> on each match:
    /// SEQUENTIAL (default), RANDOM, WEIGHTED or SWITCH.
    /// </summary>
    [JsonPropertyName("responseMode")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public ResponseMode? ResponseMode { get; set; }

    /// <summary>
    /// Relative weights index-aligned with <see cref="HttpResponses"/>; only meaningful when
    /// <see cref="ResponseMode"/> is <see cref="Models.ResponseMode.WEIGHTED"/>.
    /// </summary>
    [JsonPropertyName("responseWeights")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<int>? ResponseWeights { get; set; }

    /// <summary>
    /// Number of requests served per response block before advancing to the next; only meaningful when
    /// <see cref="ResponseMode"/> is <see cref="Models.ResponseMode.SWITCH"/> (default 1 on the server).
    /// </summary>
    [JsonPropertyName("switchAfter")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? SwitchAfter { get; set; }

    /// <summary>
    /// Cross-protocol triggers that advance a named scenario when a DNS query, WebSocket connect,
    /// gRPC request or HTTP request is observed.
    /// </summary>
    [JsonPropertyName("crossProtocolScenarios")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<CrossProtocolScenario>? CrossProtocolScenarios { get; set; }

    [JsonPropertyName("times")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public Times? Times { get; set; }

    [JsonPropertyName("timeToLive")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public TimeToLive? TimeToLive { get; set; }

    /// <summary>
    /// Optional timestamp carried so request/response log output can be round-tripped into an
    /// expectation. Preserved verbatim; not otherwise interpreted by the client.
    /// </summary>
    [JsonPropertyName("timestamp")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Timestamp { get; set; }
}
